types/tablet_v2/wlr_tablet_v2_pad.c

```diff
@@ -0,0 +1,573 @@
+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
+#include "tablet-unstable-v2-protocol.h"
+#include <assert.h>
+#include <stdlib.h>
+#include <types/wlr_tablet_v2.h>
+#include <wayland-util.h>
+#include <wlr/types/wlr_tablet_tool.h>
+#include <wlr/types/wlr_tablet_v2.h>
+#include <wlr/util/log.h>
+
+struct tablet_pad_auxiliary_user_data {
+	struct wlr_tablet_pad_client_v2 *pad;
+	size_t index;
+};
+
+void destroy_tablet_pad_v2(struct wl_resource *resource) {
+	struct wlr_tablet_pad_client_v2 *pad =
+		tablet_pad_client_from_resource(resource);
+
+	if (!pad) {
+		return;
+	}
+
+	wl_list_remove(&pad->seat_link);
+	wl_list_remove(&pad->pad_link);
+
+	/* This isn't optimal, if the client destroys the resources in another
+	 * order, it will be disconnected.
+	 * But this makes things *way* easier for us, and (untested) I doubt
+	 * clients will destroy it in another order.
+	 */
+	for (size_t i = 0; i < pad->group_count; ++i) {
+		if (pad->groups[i]) {
+			wl_resource_destroy(pad->groups[i]);
+		}
+	}
+	free(pad->groups);
+
+	for (size_t i = 0; i < pad->ring_count; ++i) {
+		if (pad->rings[i]) {
+			wl_resource_destroy(pad->rings[i]);
+		}
+	}
+	free(pad->rings);
+
+	for (size_t i = 0; i < pad->strip_count; ++i) {
+		if (pad->strips[i]) {
+			wl_resource_destroy(pad->strips[i]);
+		}
+	}
+	free(pad->strips);
+
+	free(pad);
+	wl_resource_set_user_data(resource, NULL);
+}
+
+static void handle_tablet_pad_v2_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void destroy_tablet_pad_ring_v2(struct wl_resource *resource) {
+	struct tablet_pad_auxiliary_user_data *aux = wl_resource_get_user_data(resource);
+
+	if (!aux) {
+		return;
+	}
+
+	aux->pad->rings[aux->index] = NULL;
+	free(aux);
+	wl_resource_set_user_data(resource, NULL);
+}
+
+static void handle_tablet_pad_ring_v2_set_feedback(struct wl_client *client,
+		struct wl_resource *resource, const char *description,
+		uint32_t serial) {
+	struct tablet_pad_auxiliary_user_data *aux = wl_resource_get_user_data(resource);
+	if (!aux) {
+		return;
+	}
+
+	struct wlr_tablet_v2_event_feedback evt = {
+		.serial = serial,
+		.description = description,
+		.index = aux->index
+	};
+
+	wl_signal_emit(&aux->pad->pad->events.ring_feedback, &evt);
+}
+
+static void handle_tablet_pad_ring_v2_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct zwp_tablet_pad_ring_v2_interface tablet_pad_ring_impl = {
+	.set_feedback = handle_tablet_pad_ring_v2_set_feedback,
+	.destroy = handle_tablet_pad_ring_v2_destroy,
```

Comment:
Normally, anything that begins with `handle_` is a signal handler and anything you need to distinguish to be a protocol function should be of the form `blet_pad_ring_v2_protocol_set_feedback`. This is kind of throwing me off.